### ISSUE TITLE
Standardize analysis stages' APIs

### ISSUE THREAD

**Issue by ZhangZhuoSJTU** [radareorg/radeco-lib#172]
Right now, radeco-lib have several analysis stages based on SSA (dce, sccp, cse, etc.). There are different APIs of these analyzing stages. 

- Some build a  new`Struct`, others are just a `Method`. 
- Some are passed a mutable SSA reference and change SSA directly, others are passed an immutable SSA reference and return result SSA. 

It makes both users and coders confused. We might need standardize these analyzing stages's APIs. 

For example, we might make a unified trait for the analysis. 

```rust
trait ModAnalyzer<'a> {
    fn new(rmod: &'a mut RadecoModule) -> Self;
    
    fn analyze(&mut self, analyze_all: bool);
    // when analyze_all is 
    //        true: analyze all the functions 
    //        false: only analyze matched functions
}
``` 

All the analyzers should implement the trait. And then we could open an API for RadecoModule for analysis.

```rust
fn do_analysis<T: ModAnalyzer>(&mut self, analyze_all) {
    {
        analyzer = T::new(self);
        analyzer.analyze(funcs, analyze_all);
    }
     <....Some Tail-In Work....>
}
```

This way is much more like how LLVM handle passes. And in this way, we could make analyzing parallel  if needed. 

Besides, most of analysis stages will change the structure of SSA, and some of them might destroy some invariability of SSA. Using above code, we could do some tail-in work after analyzing, like:

- verify SSA
- sort nodes' operands

**Comment by ZhangZhuoSJTU** [comment]
Any ideas? @XVilka @sushant94 @chinmaydd @kriw @HMPerson1

**Comment by chinmaydd** [comment]
@ZhangZhuoSJTU yep, that seems like a great standard to implement! It partly came up in my mind during @kriw's code review phase when there was a discrepancy in naming analysis functions (calling it `Analyzer::go`, `Analyzer::analyze`, etc.) 

If I am not getting it wrong, you mean to say that `analyze_all` should provide some logic to be run on each function SSA before choosing it for analysis?

**Comment by ZhangZhuoSJTU** [comment]
Yes, just like what you said.

I am not sure whether there will be any analysis which must be done for all the functions in the future at very beginning. Thus, I leave this logic to be able to handle this situation. (Maybe it is useless. It is ok to remove it tbh)

For now, when the user only want to analyze few functions to short working time (like what minidec does), we could simply mark these functions, and set analyze_all as false.

**Comment by sushant94** [comment]
@ZhangZhuoSJTU Yes, we should have a common trait/interface for all analysis.

But some points here, we should make the analysis of two types:
1. Ones that modify the SSA
2. Ones that only analyze and recover auxiliary information and does not modify the SSA

This would allow us to later auto-parallelize and decide which analyses can be run in parallel without issues.

Second part to that is, the level on analysis, i.e., project v/s module v/s function.  This way, function level analysis passes can be parallelized over various functions in a module, module level analysis over various modules etc. (pretty much similar to what LLVM does).

**Comment by ZhangZhuoSJTU** [comment]
@sushant94 Yea, I agree. That's a good point to start with.

Only one question, for the first type of analysis, which level should we open an API at, i.e., function(SSA) v/s module? Most analysis that modify the SSA just need information of one single function. Thus, we could add a new API for `SSAStorage` or `RadecoFunction`. However, there are also very few analysis who is inter-functions, i.e., analyzing preserved registers. Do you have some ideas about this?

**Comment by Mm7** [comment]
Hi, This should be fixed by #239

**Comment by kriw** [comment]
@Mm7 Sorry for the late reply,
> Hi, This should be fixed by #239

> * Ones that modify the SSA
> * Ones that only analyze and recover auxiliary information and does not modify the SSA

This idea is not implemented, was this discussed at https://github.com/radareorg/radeco-lib/pull/239 ?

**Comment by XVilka** [comment]
@kriw closing this one, if we have additional comments I think it is better to open a new issues.
